service/matching/version_rule_helpers.go

```diff
@@ -0,0 +1,209 @@
+	}
+	if force == true {
+		return nil
+	}
+	if hadUnfiltered && !hasUnfiltered(rules) {
+		return serviceerror.NewFailedPrecondition("update breaks requirement that at least one assignment rule must have no ramp or hint")
+	}
+	return nil
+}
+
+func countActiveAR(rules []*persistencepb.AssignmentRule) int {
+	cnt := 0
+	for _, rule := range rules {
+		if rule.DeleteTimestamp == nil {
+			cnt++
+		}
+	}
+	return cnt
+}
+
+// hasUnfiltered returns true if there exists an assignment rule with no hint, no worker ratio ramp, and no ramp percentage, or a ramp percentage of 100
+func hasUnfiltered(rules []*persistencepb.AssignmentRule) bool {
+	isUnfiltered := func(ar *taskqueue.BuildIdAssignmentRule) bool {
+		percentageRamp := ar.GetPercentageRamp()
+		return ar.GetHintFilter() == "" &&
+			ar.GetWorkerRatioRamp() == nil &&
+			(percentageRamp == nil || (percentageRamp != nil && percentageRamp.RampPercentage == 100))
+	}
+	found := false
+	for _, rule := range rules {
+		ar := rule.GetRule()
+		if rule.DeleteTimestamp == nil && isUnfiltered(ar) {
+			found = true
+		}
+	}
+	return found
+}
+
+// isInVersionSet returns true if the target build id of any assignment rule is in any of the listed version sets
+func isInVersionSet(rules []*persistencepb.AssignmentRule, sets []*persistencepb.CompatibleVersionSet) (string, bool) {
+	for _, rule := range rules {
+		ar := rule.GetRule()
+		tbid := ar.GetTargetBuildId()
+		if rule.DeleteTimestamp == nil {
+			for _, set := range sets {
+				for _, bid := range set.BuildIds {
+					if bid.GetId() == tbid {
+						return tbid, true
+					}
+				}
+			}
+		}
+	}
+	return "", false
+}
+
+// given2ActualIdx takes in the user-given index, which only counts active assignment rules, and converts it to the
+// actual index of that rule in the assignment rule list, which includes deleted rules.
+// A negative return value means index out of bounds.
+func given2ActualIdx(idx int32, rules []*persistencepb.AssignmentRule) int {
+	for i, rule := range rules {
+		if rule.DeleteTimestamp == nil {
+			if idx == 0 {
+				return i
+			}
+			idx--
+		}
+	}
+	return -1
+}
+
+func InsertAssignmentRule(timestamp *hlc.Clock,
+	data *persistencepb.VersioningData,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest_InsertBuildIdAssignmentRule,
+	maxARs int,
+	hadUnfiltered bool) (*persistencepb.VersioningData, error) {
+	if data == nil {
+		data = &persistencepb.VersioningData{AssignmentRules: make([]*persistencepb.AssignmentRule, 0)}
+	} else {
+		data = common.CloneProto(data)
+	}
+	if req.GetRuleIndex() < 0 {
+		return nil, serviceerror.NewInvalidArgument("rule index cannot be negative")
+	}
+	persistenceAR := persistencepb.AssignmentRule{
+		Rule:            req.GetRule(),
+		CreateTimestamp: timestamp,
+		DeleteTimestamp: nil,
+	}
+	rules := data.GetAssignmentRules()
+	slices.Insert(rules, given2ActualIdx(req.GetRuleIndex(), rules), &persistenceAR)
```

Comment:
in the API specification we said if the given index is too large, we insert at the end of the list, does your function do that?

Reply:
ah I missed that. thanks for catching